File: projects/03-realtime-pubsub/src/directory.rs
//! The **directory** — a persistent roster of *people* and *groups* for the
//! admin panel / playground, backed by Postgres.
//!
//! # Why this is a separate concern
//!
//! The pub/sub core (V1–V4) is deliberately **store-free**: the hub is the
//! in-memory source of truth for live subscriptions and Redis is only a *bus*,
//! never a store (see `SPEC.md` — "Redis is the bus, not the store"). Presence
//! is *soft state*: a person only exists there while their socket is open.
//!
//! This module adds the other half the admin panel needs: a **hard roster** that
//! outlives any connection. A [`Person`] here is a directory record — a name and
//! a cute emoji avatar (emoji + background color) — that exists whether or not
//! they are currently online.
//! "Online" is **not** a column you flip; it is derived at runtime from *"does
//! this person have a live WebSocket right now?"* (the presence registry). The
//! only durable *intent* we store is `autoconnect`: whether the panel should
//! bring them online on load.
//!
//! A [`Group`] is the persistent side of a topic: "Alice belongs to #eng" is a
//! [`Membership`] row that is true even when Alice is offline. Bringing her
//! online means opening a socket and `subscribe`-ing it to each of her groups.
//!
//! Keep this out of the hub/presence path — it is admin scaffolding, not a
//! vertical.
//!
//! # Your worklist
//!
//! The method bodies are [`todo!()`]. Fill them with `sqlx` — prefer the
//! compile-time-checked `query!` / `query_as!` macros. Because the scaffold has
//! **no** `query!` yet, it builds offline with no `.sqlx` cache; the moment you
//! add one you must regenerate the per-project cache (`make prepare`) or the
//! offline CI build fails. See `CLAUDE.md` → "sqlx offline cache is per-project".

use chrono::{DateTime, Utc};
use serde::Serialize;
use sqlx::PgPool;
use uuid::Uuid;

/// A directory record: someone who *can* be brought online. Persistent — this
/// outlives any WebSocket. `autoconnect` is the only online/offline state we
/// persist (an *intent*); actual online-ness is live, read from presence.
#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct Person {
    pub id: Uuid,
    pub name: String,
    /// Avatar = a chosen emoji rendered on `color` (Notion-style icon).
    pub emoji: String,
    /// Background color for the avatar (hex, e.g. `#6366f1`).
    pub color: String,
    /// Should the panel auto-connect this person on load? (Durable intent.)
    pub autoconnect: bool,
    pub created_at: DateTime<Utc>,
}

/// The persistent side of a topic. `name` *is* the topic string a socket
/// subscribes to; `emoji` + `color` are its Notion-style avatar for the panel.
#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct Group {
    pub id: Uuid,
    pub name: String,
    /// Avatar emoji for the group, rendered on `color`.
    pub emoji: String,
    pub color: String,
    pub created_at: DateTime<Utc>,
}

/// A person↔group edge. "Alice belongs to #eng", true whether or not Alice is
/// online. Bringing her online projects each edge into a live `subscribe`.
#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct Membership {
    pub person_id: Uuid,
    pub group_id: Uuid,
}

/// Handle onto the roster tables. Cheap to clone (a [`PgPool`] is `Arc` inside),
/// so it lives directly in [`AppState`](crate::AppState).
#[derive(Debug, Clone)]
pub struct Directory {
    pool: PgPool,
}

impl Directory {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    /// Borrow the pool — handy while you wire the queries up.
    pub fn pool(&self) -> &PgPool {
        &self.pool
    }

    /// Every person in the directory, newest first.
    pub async fn list_people(&self) -> Result<Vec<Person>, sqlx::Error> {
        sqlx::query_as!(Person, "SELECT * FROM people ORDER BY created_at DESC")
            .fetch_all(&self.pool)
            .await
    }

    /// Insert a new person and return the created row — the DB fills `id` and
    /// `created_at` (`INSERT ... RETURNING *`).
    pub async fn create_person(
        &self,
        name: &str,
        emoji: &str,
        color: &str,
    ) -> Result<Person, sqlx::Error> {
        sqlx::query_as!(
            Person,
            "INSERT INTO people (name, emoji, color) VALUES ($1, $2, $3) RETURNING *",
            name,
            emoji,
            color
        )
        .fetch_one(&self.pool)
        .await
    }

    /// Remove a person; their memberships cascade away via the FK.
    pub async fn delete_person(&self, id: Uuid) -> Result<(), sqlx::Error> {
        sqlx::query!("DELETE FROM people WHERE id = $1", id)
            .execute(&self.pool)
            .await
            .map(|_| ())
    }

    /// Persist the auto-connect intent for a person.
    pub async fn set_autoconnect(&self, id: Uuid, on: bool) -> Result<(), sqlx::Error> {
        sqlx::query!("UPDATE people SET autoconnect = $2 WHERE id = $1", id, on)
            .execute(&self.pool)
            .await
            .map(|_| ())
    }

    pub async fn list_groups(&self) -> Result<Vec<Group>, sqlx::Error> {
        sqlx::query_as!(Group, "SELECT * FROM groups ORDER BY created_at DESC")
            .fetch_all(&self.pool)
            .await
    }

    pub async fn create_group(
        &self,
        name: &str,
        emoji: &str,
        color: &str,
    ) -> Result<Group, sqlx::Error> {
        sqlx::query_as!(
            Group,
            "INSERT INTO groups (name, emoji, color) VALUES ($1, $2, $3) RETURNING *",
            name,
            emoji,
            color
        )
        .fetch_one(&self.pool)
        .await
    }

    pub async fn delete_group(&self, id: Uuid) -> Result<(), sqlx::Error> {
        sqlx::query!("DELETE FROM groups WHERE id = $1", id)
            .execute(&self.pool)
            .await
            .map(|_| ())
    }

    /// Every edge. The panel joins these against people+groups client-side to
    /// render "who is in what".
    pub async fn memberships(&self) -> Result<Vec<Membership>, sqlx::Error> {
        sqlx::query_as!(Membership, "SELECT person_id, group_id FROM memberships")
            .fetch_all(&self.pool)
            .await
    }

    /// Add a person to a group. Make it idempotent (`ON CONFLICT DO NOTHING`) so
    /// re-adding an existing member isn't an error.
    pub async fn add_member(&self, person_id: Uuid, group_id: Uuid) -> Result<(), sqlx::Error> {
        sqlx::query!(
            "INSERT INTO memberships (person_id, group_id) VALUES ($1, $2) ON CONFLICT DO NOTHING",
            person_id,
            group_id
        )
        .execute(&self.pool)
        .await
        .map(|_| ())
    }

    pub async fn remove_member(&self, person_id: Uuid, group_id: Uuid) -> Result<(), sqlx::Error> {
        sqlx::query!(
            "DELETE FROM memberships WHERE person_id = $1 AND group_id = $2",
            person_id,
            group_id
        )
        .execute(&self.pool)
        .await
        .map(|_| ())
    }
}

#[cfg(test)]
mod tests {
    //! Directory integration tests. `#[sqlx::test]` gives each test its own
    //! freshly-migrated Postgres database (needs `DATABASE_URL`).

    use super::*;

    fn dir(pool: PgPool) -> Directory {
        Directory::new(pool)
    }

    #[sqlx::test]
    async fn create_person_returns_row_with_defaults(pool: PgPool) {
        let d = dir(pool);
        let p = d
            .create_person("Alice", "🧘", "#6366f1")
            .await
            .expect("create");

        assert_eq!(p.name, "Alice");
        assert_eq!(p.emoji, "🧘");
        assert_eq!(p.color, "#6366f1");
        assert!(!p.autoconnect, "autoconnect defaults to false");
        assert!(!p.id.is_nil());
    }

    #[sqlx::test]
    async fn list_people_returns_newest_first(pool: PgPool) {
        let d = dir(pool);
        let older = d.create_person("Older", "👴", "#111111").await.unwrap();
        let newer = d.create_person("Newer", "👶", "#222222").await.unwrap();

        let people = d.list_people().await.expect("list");
        assert_eq!(people.len(), 2);
        assert_eq!(people[0].id, newer.id);
        assert_eq!(people[1].id, older.id);
        assert!(people[0].created_at >= people[1].created_at);
    }

    #[sqlx::test]
    async fn delete_person_removes_row(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Gone", "👋", "#000000").await.unwrap();

        d.delete_person(p.id).await.expect("delete");

        let people = d.list_people().await.unwrap();
        assert!(people.is_empty());
    }

    #[sqlx::test]
    async fn set_autoconnect_persists_intent(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Bob", "🤖", "#ff0000").await.unwrap();
        assert!(!p.autoconnect);

        d.set_autoconnect(p.id, true).await.expect("set on");
        let on = d
            .list_people()
            .await
            .unwrap()
            .into_iter()
            .find(|x| x.id == p.id)
            .expect("person still listed");
        assert!(on.autoconnect);

        d.set_autoconnect(p.id, false).await.expect("set off");
        let off = d
            .list_people()
            .await
            .unwrap()
            .into_iter()
            .find(|x| x.id == p.id)
            .unwrap();
        assert!(!off.autoconnect);
    }

    #[sqlx::test]
    async fn create_group_returns_row(pool: PgPool) {
        let d = dir(pool);
        let g = d
            .create_group("eng", "🎨", "#10b981")
            .await
            .expect("create");

        assert_eq!(g.name, "eng");
        assert_eq!(g.emoji, "🎨");
        assert_eq!(g.color, "#10b981");
        assert!(!g.id.is_nil());
    }

    #[sqlx::test]
    async fn list_groups_returns_newest_first(pool: PgPool) {
        let d = dir(pool);
        let older = d.create_group("alpha", "🅰️", "#111111").await.unwrap();
        let newer = d.create_group("beta", "🅱️", "#222222").await.unwrap();

        let groups = d.list_groups().await.expect("list");
        assert_eq!(groups.len(), 2);
        assert_eq!(groups[0].id, newer.id);
        assert_eq!(groups[1].id, older.id);
    }

    #[sqlx::test]
    async fn delete_group_removes_row(pool: PgPool) {
        let d = dir(pool);
        let g = d.create_group("temp", "🗑️", "#000000").await.unwrap();

        d.delete_group(g.id).await.expect("delete");

        assert!(d.list_groups().await.unwrap().is_empty());
    }

    #[sqlx::test]
    async fn add_member_and_list_memberships(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Alice", "🧘", "#6366f1").await.unwrap();
        let g = d.create_group("eng", "🎨", "#10b981").await.unwrap();

        d.add_member(p.id, g.id).await.expect("add");

        let edges = d.memberships().await.expect("memberships");
        assert_eq!(edges.len(), 1);
        assert_eq!(edges[0].person_id, p.id);
        assert_eq!(edges[0].group_id, g.id);
    }

    #[sqlx::test]
    async fn add_member_is_idempotent(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Alice", "🧘", "#6366f1").await.unwrap();
        let g = d.create_group("eng", "🎨", "#10b981").await.unwrap();

        d.add_member(p.id, g.id).await.expect("first add");
        d.add_member(p.id, g.id)
            .await
            .expect("second add must not error");

        assert_eq!(d.memberships().await.unwrap().len(), 1);
    }

    #[sqlx::test]
    async fn remove_member_deletes_edge(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Alice", "🧘", "#6366f1").await.unwrap();
        let g = d.create_group("eng", "🎨", "#10b981").await.unwrap();
        d.add_member(p.id, g.id).await.unwrap();

        d.remove_member(p.id, g.id).await.expect("remove");

        assert!(d.memberships().await.unwrap().is_empty());
    }

    #[sqlx::test]
    async fn delete_person_cascades_memberships(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Alice", "🧘", "#6366f1").await.unwrap();
        let g = d.create_group("eng", "🎨", "#10b981").await.unwrap();
        d.add_member(p.id, g.id).await.unwrap();

        d.delete_person(p.id).await.expect("delete person");

        assert!(d.memberships().await.unwrap().is_empty());
        assert_eq!(d.list_groups().await.unwrap().len(), 1, "group survives");
    }

    #[sqlx::test]
    async fn delete_group_cascades_memberships(pool: PgPool) {
        let d = dir(pool);
        let p = d.create_person("Alice", "🧘", "#6366f1").await.unwrap();
        let g = d.create_group("eng", "🎨", "#10b981").await.unwrap();
        d.add_member(p.id, g.id).await.unwrap();

        d.delete_group(g.id).await.expect("delete group");

        assert!(d.memberships().await.unwrap().is_empty());
        assert_eq!(d.list_people().await.unwrap().len(), 1, "person survives");
    }
}

File: projects/03-realtime-pubsub/src/realtime_pubsub/directory.py
"""The **directory** — a persistent roster of *people* and *groups* for the
admin panel / playground, backed by Postgres.

# Why this is a separate concern

The pub/sub core (V1-V4) is deliberately **store-free**: the hub is the
in-memory source of truth for live subscriptions and Redis is only a *bus*,
never a store (see `SPEC.md` — "Redis is the bus, not the store"). Presence is
*soft state*: a person only exists there while their socket is open.

This module adds the other half the admin panel needs: a **hard roster** that
outlives any connection. A `Person` here is a directory record — a name and a
cute emoji avatar — that exists whether or not they are currently online.
"Online" is **not** a column you flip; it is derived at runtime from *"does this
person have a live WebSocket right now?"* (the presence registry). The only
durable *intent* stored is `autoconnect`: whether the panel should bring them
online on load.

A `Group` is the persistent side of a topic: "Alice belongs to #eng" is a
`Membership` row that is true even when Alice is offline. Bringing her online
means opening a socket and subscribing it to each of her groups.

Keep this out of the hub/presence path — it is admin scaffolding, not a vertical.

# Shape

Two type families on purpose, which is the standard FastAPI + SQLAlchemy split
and worth having in the fingers:

* `PersonRow` / `GroupRow` / `MembershipRow` are the **table** definitions
  (SQLAlchemy 2.0 declarative, `Mapped[...]` annotations).
* `Person` / `Group` / `Membership` are the **wire** models (pydantic), built
  from a row with `model_validate`. They are what the API returns, so a column
  added to a table is not automatically published to clients.

The schema itself still lives in `migrations/0001_directory.sql` — the same file
the Rust version used. It is applied at startup rather than generated from these
classes, so the SQL stays the single source of truth and the two cannot drift.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict
from sqlalchemy import Boolean, DateTime, ForeignKey, String, delete, select, text
from sqlalchemy.dialects.postgresql import UUID as PgUUID
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncEngine, async_sessionmaker, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

__all__ = [
    "Directory",
    "Group",
    "Membership",
    "Person",
    "connect",
]


class Base(DeclarativeBase):
    pass


class PersonRow(Base):
    __tablename__ = "people"

    id: Mapped[uuid.UUID] = mapped_column(
        PgUUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    name: Mapped[str] = mapped_column(String, nullable=False)
    emoji: Mapped[str] = mapped_column(String, nullable=False, server_default="🧘")
    color: Mapped[str] = mapped_column(String, nullable=False, server_default="#6366f1")
    autoconnect: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("now()")
    )


class GroupRow(Base):
    __tablename__ = "groups"

    id: Mapped[uuid.UUID] = mapped_column(
        PgUUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    name: Mapped[str] = mapped_column(String, nullable=False, unique=True)
    emoji: Mapped[str] = mapped_column(String, nullable=False, server_default="🎨")
    color: Mapped[str] = mapped_column(String, nullable=False, server_default="#6366f1")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("now()")
    )


class MembershipRow(Base):
    __tablename__ = "memberships"

    person_id: Mapped[uuid.UUID] = mapped_column(
        PgUUID(as_uuid=True), ForeignKey("people.id", ondelete="CASCADE"), primary_key=True
    )
    group_id: Mapped[uuid.UUID] = mapped_column(
        PgUUID(as_uuid=True), ForeignKey("groups.id", ondelete="CASCADE"), primary_key=True
    )
    joined_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("now()")
    )


# --- wire models --------------------------------------------------------------


class Person(BaseModel):
    """A directory record: someone who *can* be brought online."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    emoji: str
    """Avatar = a chosen emoji rendered on `color` (Notion-style icon)."""
    color: str
    autoconnect: bool
    """Should the panel auto-connect this person on load? (Durable intent.)"""
    created_at: datetime


class Group(BaseModel):
    """The persistent side of a topic. `name` *is* the topic string a socket
    subscribes to."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    emoji: str
    color: str
    created_at: datetime


class Membership(BaseModel):
    """A person-group edge, true whether or not the person is online."""

    model_config = ConfigDict(from_attributes=True)

    person_id: uuid.UUID
    group_id: uuid.UUID


# --- the handle ---------------------------------------------------------------


class Directory:
    """Handle onto the roster tables."""

    __slots__ = ("_engine", "_session")

    def __init__(self, engine: AsyncEngine) -> None:
        self._engine = engine
        # `expire_on_commit=False` so the attributes of a returned row are still
        # readable after the session's commit. Without it, reading `person.name`
        # on the way out triggers a *lazy refresh* — a second round trip, and in
        # async SQLAlchemy an error rather than a silent one.
        self._session = async_sessionmaker(engine, expire_on_commit=False)

    @property
    def engine(self) -> AsyncEngine:
        """The underlying engine — used by the `/debug/health` probe."""
        return self._engine

    async def list_people(self) -> list[Person]:
        """Every person in the directory, newest first."""
        async with self._session() as session:
            rows = await session.scalars(select(PersonRow).order_by(PersonRow.created_at.desc()))
            return [Person.model_validate(row) for row in rows]

    async def create_person(self, name: str, emoji: str, color: str) -> Person:
        """Insert a new person and return the created row — the DB fills `id`
        and `created_at`."""
        async with self._session.begin() as session:
            row = PersonRow(name=name, emoji=emoji, color=color)
            session.add(row)
            # Force the INSERT now so the server defaults come back and the
            # object is fully populated before the session closes.
            await session.flush()
            await session.refresh(row)
            return Person.model_validate(row)

    async def delete_person(self, person_id: uuid.UUID) -> None:
        """Remove a person; their memberships cascade away via the FK."""
        async with self._session.begin() as session:
            await session.execute(delete(PersonRow).where(PersonRow.id == person_id))

    async def set_autoconnect(self, person_id: uuid.UUID, on: bool) -> None:
        """Persist the auto-connect intent for a person."""
        async with self._session.begin() as session:
            row = await session.get(PersonRow, person_id)
            if row is not None:
                row.autoconnect = on

    async def list_groups(self) -> list[Group]:
        async with self._session() as session:
            rows = await session.scalars(select(GroupRow).order_by(GroupRow.created_at.desc()))
            return [Group.model_validate(row) for row in rows]

    async def create_group(self, name: str, emoji: str, color: str) -> Group:
        async with self._session.begin() as session:
            row = GroupRow(name=name, emoji=emoji, color=color)
            session.add(row)
            await session.flush()
            await session.refresh(row)
            return Group.model_validate(row)

    async def delete_group(self, group_id: uuid.UUID) -> None:
        async with self._session.begin() as session:
            await session.execute(delete(GroupRow).where(GroupRow.id == group_id))

    async def memberships(self) -> list[Membership]:
        """Every edge. The panel joins these against people + groups client-side
        to render "who is in what"."""
        async with self._session() as session:
            rows = await session.scalars(select(MembershipRow))
            return [Membership.model_validate(row) for row in rows]

    async def add_member(self, person_id: uuid.UUID, group_id: uuid.UUID) -> None:
        """Add a person to a group.

        Idempotent via `ON CONFLICT DO NOTHING`, so re-adding an existing member
        is not an error — the panel fires this on every toggle and should not
        have to check first.
        """
        async with self._session.begin() as session:
            statement = (
                pg_insert(MembershipRow)
                .values(person_id=person_id, group_id=group_id)
                .on_conflict_do_nothing(index_elements=["person_id", "group_id"])
            )
            await session.execute(statement)

    async def remove_member(self, person_id: uuid.UUID, group_id: uuid.UUID) -> None:
        async with self._session.begin() as session:
            await session.execute(
                delete(MembershipRow).where(
                    MembershipRow.person_id == person_id,
                    MembershipRow.group_id == group_id,
                )
            )

    async def aclose(self) -> None:
        await self._engine.dispose()


def _statements(schema_sql: str) -> list[str]:
    """Split a DDL file into individual statements.

    asyncpg sends everything through the *extended* query protocol, which
    accepts exactly one statement per call — hand it a whole `.sql` file and it
    raises "cannot insert multiple commands into a prepared statement". So the
    file is split here. Comments are stripped first, because a `--` line could
    otherwise contribute a stray semicolon and produce an empty statement.
    """
    without_comments = "\n".join(
        line for line in schema_sql.splitlines() if not line.lstrip().startswith("--")
    )
    return [stmt.strip() for stmt in without_comments.split(";") if stmt.strip()]


async def connect(database_url: str, schema_sql: str) -> Directory:
    """Open the pool and apply the schema.

    `schema_sql` is `migrations/0001_directory.sql`, which is idempotent
    (`CREATE TABLE IF NOT EXISTS`), so running it on every boot is safe and
    keeps the `.sql` file authoritative rather than these ORM classes.

    `pool_size` is a knob the SPEC asks you to set deliberately rather than
    inherit: it caps how many concurrent admin queries can be in flight, and it
    wants to be reasoned about next to how many uvicorn workers you run — five
    connections per worker times eight workers is forty against a Postgres whose
    default `max_connections` is a hundred.
    """
    engine = create_async_engine(
        _asyncpg_url(database_url),
        pool_size=5,
        pool_pre_ping=True,
    )
    async with engine.begin() as conn:
        for statement in _statements(schema_sql):
            await conn.execute(text(statement))
    return Directory(engine)


def _asyncpg_url(database_url: str) -> str:
    """Force the asyncpg driver onto a plain Postgres URL.

    The same `DATABASE_URL` is shared with docker-compose and the Rust-era
    tooling, both of which want `postgres://`. SQLAlchemy needs the driver named
    in the scheme, and picking it here means the `.env` does not have to carry a
    Python-specific URL.
    """
    for prefix in ("postgresql+asyncpg://", "postgres://", "postgresql://"):
        if database_url.startswith(prefix):
            return "postgresql+asyncpg://" + database_url[len(prefix) :]
    return database_url
